pkg/cmd/controller/controller_build.go

```diff
@@ -195,56 +194,40 @@ func (o *ControllerBuildOptions) Run() error {
 		return errors.Wrap(err, "failed to label the PipelineActivity resources")
 	}
 
+	var handlers watcher.HandlerFuncs
 	if tektonEnabled {
-		pod := &corev1.Pod{}
-		log.Logger().Infof("Watching for Pods in namespace %s", util.ColorInfo(ns))
-		listWatch := cache.NewListWatchFromClient(kubeClient.CoreV1().RESTClient(), "pods", ns, fields.Everything())
-		kube.SortListWatchByName(listWatch)
-		_, controller := cache.NewInformer(
-			listWatch,
-			pod,
-			time.Minute*10,
-			cache.ResourceEventHandlerFuncs{
-				AddFunc: func(obj interface{}) {
-					o.onPipelinePod(obj, kubeClient, jxClient, tektonClient, ns)
-				},
-				UpdateFunc: func(oldObj, newObj interface{}) {
-					o.onPipelinePod(newObj, kubeClient, jxClient, tektonClient, ns)
-				},
-				DeleteFunc: func(obj interface{}) {
-				},
+		log.Logger().Infof("Watching for tekton Pods in namespace %s", util.ColorInfo(ns))
+		handlers = watcher.HandlerFuncs{
+			AddFunc: func(obj runtime.Object) {
+				o.onPipelinePod(obj, kubeClient, jxClient, tektonClient, ns)
 			},
-		)
-
-		stop := make(chan struct{})
-		go controller.Run(stop)
+			ModifyFunc: func(obj runtime.Object) {
+				o.onPipelinePod(obj, kubeClient, jxClient, tektonClient, ns)
+			},
+			DeleteFunc: func(obj runtime.Object) {
+			},
+		}
 	} else {
-		pod := &corev1.Pod{}
 		log.Logger().Infof("Watching for Knative build pods in namespace %s", util.ColorInfo(ns))
-		listWatch := cache.NewListWatchFromClient(kubeClient.CoreV1().RESTClient(), "pods", ns, fields.Everything())
-		kube.SortListWatchByName(listWatch)
-		_, controller := cache.NewInformer(
-			listWatch,
-			pod,
-			time.Minute*10,
-			cache.ResourceEventHandlerFuncs{
-				AddFunc: func(obj interface{}) {
-					o.onPod(obj, kubeClient, jxClient, ns)
-				},
-				UpdateFunc: func(oldObj, newObj interface{}) {
-					o.onPod(newObj, kubeClient, jxClient, ns)
-				},
-				DeleteFunc: func(obj interface{}) {
-				},
+		handlers = watcher.HandlerFuncs{
+			AddFunc: func(obj runtime.Object) {
+				o.onPod(obj, kubeClient, jxClient, ns)
 			},
-		)
-
-		stop := make(chan struct{})
-		go controller.Run(stop)
+			ModifyFunc: func(obj runtime.Object) {
+				o.onPod(obj, kubeClient, jxClient, ns)
+			},
+			DeleteFunc: func(obj runtime.Object) {
+			},
+		}
 	}
 
-	// Wait forever
-	select {}
+	watcher := watcher.NewPodWatcher(kubeClient, ns, metav1.ListOptions{}, handlers)
```

Comment:
OOI, apart of abstracting some of the watching logic into `PodWatcher`, how does this all fix #5395? What was the underlying issue?

Reply:
the old code wrapped all of the handling watch errors, so if a watcher failed the pod just sat there forever, never attempting a reconnect.

now we break the run loop if a watch fails: https://github.com/jenkins-x/jx/blob/d7f746d25566ac9d0e7a469083702fd883aeae51/pkg/kube/watcher/watcher.go#L65-L68

which we then recreate everything again - or if the list/watch fail we terminate the pod (e.g. if the API server is down due to an upgrade of k8s versions)